packages/aggregator/src/connections/DataSourceEntityProvider.js

```diff
@@ -0,0 +1,52 @@
+/*
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ *
+ */
+
+export class DataSourceEntityProvider {
+  constructor(entityApi) {
+    this.entityApi = entityApi;
+  }
+
+  parseEntityType(entityType) {
+    return entityType == 'requested' ? undefined : entityType;
+  }
+
+  async getDataSourceEntities(
+    hierarchyName,
+    entityCodes,
+    dataSourceEntityType,
+    dataSourceEntityFilter,
+  ) {
+    return this.entityApi.getRelativesOfEntities(hierarchyName, entityCodes, {
+      field: 'code',
+      filter: {
+        type: this.parseEntityType(dataSourceEntityType),
+        ...dataSourceEntityFilter,
+      },
+    });
+  }
+
+  async getDataSourceEntitiesAndRelationships(
+    hierarchyName,
+    entityCodes,
+    aggregationEntityType,
+    dataSourceEntityType,
+    dataSourceEntityFilter,
+  ) {
+    const response = await this.entityApi.getRelationshipsOfEntitiesByDescendant(
+      hierarchyName,
+      entityCodes,
+      { field: 'code' },
+      { filter: { type: this.parseEntityType(aggregationEntityType) } },
+      { filter: { type: this.parseEntityType(dataSourceEntityType), ...dataSourceEntityFilter } },
+    );
```

Comment:
Gotta say, I am so happy with how clean this interface is 😄 Reads so nicely to me!